lib/makara/context.rb

```diff
@@ -1,83 +1,159 @@
 require 'digest/md5'
 
-# Keeps track of the current and previous context (hexdigests)
-# If a new context is needed it can be generated via Makara::Context.generate
-
+# Keeps track of the current stickiness state for different Makara configurations
 module Makara
   class Context
-    class << self
+    attr_accessor :stored_data, :staged_data
+
+    def initialize(context_data)
+      @stored_data = context_data
+      @staged_data = {}
+      @dirty = @was_dirty = false
+    end
+
+    def stage(proxy_id, ttl)
+      staged_data[proxy_id] = ttl.to_f
+    end
+
+    def stuck?(proxy_id)
+      stored_data[proxy_id] && !expired?(stored_data[proxy_id])
+    end
+
+    def staged?(proxy_id)
+      staged_data.key?(proxy_id)
+    end
 
-      def generate(seed = nil)
-        seed ||= "#{Time.now.to_i}#{Thread.current.object_id}#{rand(99999)}"
-        Digest::MD5.hexdigest(seed)
+    def release(proxy_id)
+      @dirty ||= !!stored_data.delete(proxy_id)
+      staged_data.delete(proxy_id)
+    end
+
+    def release_all
+      if self.stored_data.any?
+        self.stored_data = {}
+        # We need to track a change made to the current stored data
+        # so we can commit it later
+        @dirty = true
       end
+      self.staged_data = {}
+    end
+
+    # Stores the staged data with an expiration time based on the current time,
+    # and clears any expired entries. Returns true if any changes were made to
+    # the current store
+    def commit
+      release_expired
+      store_staged_data
```

Comment:
Hmm... the `0` values won't get written in any case, since we'll only store `proxy_id, ttl` pairs with `ttl > 0`. I don't see any other reason why we should check for expiration of data staged in this request, in theory no staged data should be expired because we'll be comparing `Time.now + ttl` (with `ttl > 0`) with `Time.now`, so I believe this is the right order. Am I missing anything?